### PR TITLE
feat: add `DynamicDoryEvaluationProof`

### PR DESCRIPTION
**Pull request by JayWhite2357**
# Rationale for this change

This is the final PR to add the dynamic-sized Dory PCS.

# What changes are included in this PR?

* dynamic_dory_helper module and functions are added
* The `DynamicDoryEvaluationProof` type is added
* Tests are added

To best review this PR, look at the individual commits.
* The first 3 commits: https://github.com/spaceandtimelabs/sxt-proof-of-sql/pull/151/files/44de1aea8d011cbeb1fa9da9aeeabcf1ff87ef50 are copying code from the existing non-dynamic scheme
* The second 3 commits are the actual new code. This is predominantly refactoring the copied files. https://github.com/spaceandtimelabs/sxt-proof-of-sql/pull/151/files/44de1aea8d011cbeb1fa9da9aeeabcf1ff87ef50..9e8b62f577be1561f8dab64534542ede99975482

# Are these changes tested?

Yes

### CODE REVIEW

File: crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_helper.rs [hunk 1]
```diff
@@ -1,36 +1,214 @@
...
+    for (i, v) in a.iter().enumerate() {
+        let (row, column) = row_and_column_from_index(i);
+        v_vec[column] += hi_vec[row] * v;
+    }
+    v_vec
 }
 
 /// Compute the commitments to the rows of the matrix M that is derived from `a`.
-pub(super) fn compute_T_vec_prime(
+pub(super) fn compute_dynamic_T_vec_prime(
```
Comment (jacobtrombetta) [inline crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_helper.rs:21]: Is there a test for this function?

File: crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_helper.rs [hunk 2]
```diff
@@ -1,36 +1,214 @@
...
+    ProverSetup, F,
+};
+use crate::proof_primitive::dory::dynamic_dory_standard_basis_helper::compute_dynamic_standard_basis_vecs;
+use ark_ff::Field;
+use itertools::Itertools;
+
 /// Compute the evaluations of the columns of the matrix M that is derived from `a`.
-pub(super) fn compute_v_vec(a: &[F], L_vec: &[F], sigma: usize, nu: usize) -> Vec<F> {
-    todo!()
+pub(super) fn compute_dynamic_v_vec(a: &[F], hi_vec: &[F], nu: usize) -> Vec<F> {
```
Comment (jacobtrombetta) [inline crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_helper.rs:11]: Is there a test for this function?

File: crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_cpu.rs [hunk 3]
```diff
@@ -1,55 +1,39 @@
...
-            [rows_offset..(rows_offset + remaining_row_count + 1)],
+    let Gamma_1 = setup.Gamma_1.last().unwrap();
+    let Gamma_2 = setup.Gamma_2.last().unwrap();
+    let (first_row, _) = row_and_column_from_index(offset);
+    let (last_row, _) = row_and_column_from_index(offset + column.len() - 1);
+    let mut row_commits = vec![G1Projective::from(G1Affine::identity()); last_row - first_row + 1];
+    for (i, v) in column.iter().enumerate() {
+        let (row, col) = row_and_column_from_index(i + offset);
+        row_commits[row - first_row] += Gamma_1[col] * v.into().0;
+    }
```
Comment (tlovell-sxt) [inline crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_cpu.rs:25]: NIT: more optional fold
```suggestion
    let row_commits = column
        .iter()
        .enumerate()
        .fold(vec![G1Projective::from(G1Affine::identity()); last_row - first_row + 1], |mut row_commits, (i, v)| {
            let (row, col) = row_and_column_from_index(i + offset);
            row_commits[row - first_row] += Gamma_1[col] * v.into().0;
            row_commits
        });
```

**Review by tlovell-sxt — APPROVED** [review]
just NITs, no worries if you disagree about folding, I just like to minimize the mutability

File: crates/proof-of-sql/src/proof_primitive/dory/mod.rs [hunk 4]
```diff
@@ -143,5 +143,15 @@ mod pack_scalars;
...
+mod dynamic_dory_helper;
 mod dynamic_dory_standard_basis_helper;
 mod dynamic_dory_structure;
+use dynamic_dory_commitment_helper_cpu::compute_dynamic_dory_commitments;
+mod dynamic_dory_commitment;
+mod dynamic_dory_commitment_evaluation_proof;
+pub use dynamic_dory_commitment::DynamicDoryCommitment;
+#[cfg(test)]
+mod dynamic_dory_commitment_evaluation_proof_test;
+pub use dynamic_dory_commitment_evaluation_proof::DynamicDoryEvaluationProof;
```
[the remark this answers is not in the document]
Comment (JayWhite2357) [reply crates/proof-of-sql/src/proof_primitive/dory/mod.rs:157]: Not really. `rust fmt` automatically does some type of ordering it as long as the lines don't have a blank space between then. Feels a bit random, though.

File: crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_cpu.rs [hunk 5]
```diff
@@ -1,55 +1,39 @@
...
-    // Compute the commitment for the entire matrix.
-    DoryCommitment(pairings::multi_pairing(
-        once(first_row_commit).chain(remaining_row_commits),
-        &setup.prover_setup().Gamma_2.last().unwrap()
-            [rows_offset..(rows_offset + remaining_row_count + 1)],
+    let Gamma_1 = setup.Gamma_1.last().unwrap();
+    let Gamma_2 = setup.Gamma_2.last().unwrap();
+    let (first_row, _) = row_and_column_from_index(offset);
+    let (last_row, _) = row_and_column_from_index(offset + column.len() - 1);
+    let mut row_commits = vec![G1Projective::from(G1Affine::identity()); last_row - first_row + 1];
```
[the remark this answers is not in the document]
Comment (JayWhite2357) [reply crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_cpu.rs:21]: If I try that I get:
> no function or associated item named `identity` found for struct `ark_ec::short_weierstrass::Projective` in the current scope
function or associated item not found in `Projective<Config>`

File: crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_helper.rs [hunk 6]
```diff
@@ -1,36 +1,214 @@
...
+    let mut v_vec = vec![F::ZERO; 1 << nu];
+    for (i, v) in a.iter().enumerate() {
+        let (row, column) = row_and_column_from_index(i);
+        v_vec[column] += hi_vec[row] * v;
+    }
+    v_vec
 }
```
[the remark this answers is not in the document]
Comment (JayWhite2357) [reply crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_helper.rs]: Added some docs about this that hopefully clarify.

**Review by tlovell-sxt — APPROVED** [review]

**Pull request merged**